pkg/controller/secretsyncer/secretsyncer.go

```diff
@@ -0,0 +1,205 @@
+	allTargetClusters := sets.NewString(referenceClusterName)
+	for targetCluster, targetClusterManager := range otherBuildClusters {
+		r.clients[targetCluster] = targetClusterManager.GetClient()
+		allTargetClusters.Insert(targetCluster)
+	}
+
+	if err := c.Watch(
+		source.NewKindWithCache(&corev1.Secret{}, referenceCluster.GetCache()),
+		&handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(
+			func(mo handler.MapObject) []reconcile.Request {
+				// The reference cluster enq all clusters
+				var requests []reconcile.Request
+				for _, targetCluster := range allTargetClusters.List() {
+					requests = append(requests, requestForCluster(targetCluster, mo.Meta.GetNamespace(), mo.Meta.GetName()))
+				}
+				return requests
+			},
+		)},
+	); err != nil {
+		return fmt.Errorf("failed to create watch on reference cluster %s: %w", referenceClusterName, err)
+	}
+
+	return nil
+}
+
+type reconciler struct {
+	ctx                    context.Context
+	log                    *logrus.Entry
+	config                 config.Getter
+	referenceClusterClient ctrlruntimeclient.Client
+	clients                map[string]ctrlruntimeclient.Client
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", request.String())
+	err := r.reconcile(log, request)
+	if err != nil {
+		log.WithError(err).Error("Reconciliation failed")
+	} else {
+		log.Trace("Finished reconciliation successfully")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(log *logrus.Entry, encodedRequest reconcile.Request) error {
+	cluster, sourceSecretName, err := decodeRequest(encodedRequest)
+	if err != nil {
+		return fmt.Errorf("failed to decode request %s: %w", encodedRequest, err)
+	}
+	*log = *log.WithField("cluster", cluster)
+	client, ok := r.clients[cluster]
+	if !ok {
+		return controllerutil.TerminalError(fmt.Errorf("no client for cluster %s available", cluster))
+	}
+	log.WithField("request", sourceSecretName.String()).Trace("Reconciling")
+
+	source := &corev1.Secret{}
+	if err := r.referenceClusterClient.Get(r.ctx, sourceSecretName, source); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get secret %s: %w", sourceSecretName.String(), err)
+	}
+	if source.DeletionTimestamp != nil {
+		log.Debug("not doing work for secret because it is being deleted")
+		return nil
+	}
+
+	if len(source.Data) == 0 {
+		log.Trace("not updating target secret as source has no data")
+		return nil
+	}
+
+	var mirrorErrors []error
+	for _, mirrorConfig := range r.config().Secrets {
+		if mirrorConfig.From.Namespace != sourceSecretName.Namespace || mirrorConfig.From.Name != sourceSecretName.Name {
+			continue
+		}
+		// TODO: Ideally we would would create one reconcile.Request per target. This here means
+		// that if a single target is broken, it will send all targets into exponential backoff.
```

Comment:
> This corresponds If a secret changes in the source namespace, a handler that creates a reconcile.Request per cluster needs to be used, right?

No, that is implemented here: https://github.com/openshift/ci-tools/pull/1077/files#diff-eaff0abece754b19eedc4a83fb0b03bbR65

The comment refers to the fact that a secret could be configured to get mirrored into more than one target secret (not sure if that case actually exists, but the code would handle it). Now if for example three is a config to mirror into `namespace-a/secret-a` and one to mirror into `namespace-b/secret-b` but `namespace-b` doesn't exist, the second part of it will always fail, because its not possible to create something in a namespace that doesn't exist.

Failing here means that ultimatively the `Reconcile` function will return an error. This causes the underlying machinery to re-enq the same request with an exponential backoff. On every retry, it will still fail because that namespace doesn't exist. This means that if `secret-a` which we can deploy changes, it might only get deployed after a long delay because all of the retrying even though that delay is because `secret-b` has a problem.

Reply:
Ack.